src/interval_domain.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+//
+
+use expression::ExpressionType::{self, *};
+use std::cmp;
+use std::convert::TryFrom;
+
+/// An element of the Interval domain is a range of i128 numbers denoted by a lower bound and
+/// upper bound. A lower bound of std::i128::MIN denotes -infinity and an upper bound of
+/// std::i128::MAX denotes +infinity.
+/// Interval domain elements are constructed on demand from AbstractDomain expressions.
+/// They are most useful for checking if an array index is within bounds.
+#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialOrd, PartialEq, Hash)]
+pub struct IntervalDomain {
```

Comment:
It seems like
 
  type Value = Min | Number i128 | Max;
  type Type =  Top | { lower_bound: Value, upper_bound: Value } | Top

is more 'correct', but using things like MIN and MAX for Value and lower_bound: 1, upper_bound: 0 for bottom is just to make operations more convenient?

Reply:
Once you introduce more cases, everything pretty much explodes.